Refactor issueShark to adapt pycoShark release 3.0

myousfi96 opened the pull request:
Changes 

- Github backend
- Jira backend
- Bugzilla backed.
- bases backend

atrautsch commented on issueshark/backends/basebackend.py:
```diff
@@ -18,17 +18,22 @@ def identifier(self):
         self.config = cfg
         self.issue_system_id = issue_system_id
         self.project_id = project_id
+        self.last_systme_id = last_system_id
```
last_systme_id ->last_system_id

atrautsch commented on issueshark/backends/github.py:
```diff
@@ -108,36 +103,41 @@ def store_issue(self, raw_issue):
             # We can not return here, as the issue might be updated. This means, that the title could be updated
             # as well as comments and new events
-            issue = Issue.objects(issue_system_id=self.issue_system_id, external_id=str(raw_issue['number'])).get()
+            mongo_issue = Issue.objects(issue_system_ids=self.last_systme_id, external_id=self.issue_id).get()
```
last_systme_id

atrautsch commented on issueshark/backends/github.py:
```diff
@@ -157,38 +157,29 @@ def _process_events(self, system_id, mongo_issue):
+            mongo_event = None
+            if mongo_issue:
+                try:
+                    mongo_event = IssueEvent.objects(external_id=str(raw_event['id']), issue_id=mongo_issue.id).get()
```
Could this return multiple issue events for multiple issue_system_ids if we have updated the data?

atrautsch commented on issueshark/backends/github.py:
```diff
@@ -157,38 +157,29 @@ def _process_events(self, system_id, mongo_issue):
+                except DoesNotExist:
+                    mongo_event = None
+
+            new_event = IssueEvent(external_id=str(raw_event['id']), created_at=created_at, status=raw_event['event'])
```
How is this linked to the issue?

myousfi96 replied:
I linked it in the save_issues. function I added the issue ID to it.

atrautsch commented on issueshark/backends/github.py:
```diff
@@ -226,42 +217,41 @@ def _set_old_and_new_value_for_event(self, event, raw_event, mongo_issue):
+                except DoesNotExist:
+                    mongo_comment = None
+
+            new_comment = IssueComment(
```
How is this linked to the issue?

myousfi96 replied:
I linked it in the save_issues. function I added the issue ID to it.

atrautsch commented on issueshark/backends/jirabackend.py:
```diff
@@ -241,41 +229,42 @@ def _store_comments(self, jira_issue, mongo_issue_id):
+            mongo_comment = None
+            if mongo_issue:
+                try:
+                    mongo_comment = IssueComment.objects(external_id=comment.id, issue_id=mongo_issue.id).get()
```
What if an issue or part of an issue was changed and we update the data, could this return multiple issue comments?

myousfi96 replied:
No, because Issue id is unique for every time we update the data. For example, if we have updated data we will create a new issue id therefore there is no problem here.

atrautsch commented on issueshark/backends/jirabackend.py:
```diff
@@ -418,16 +409,17 @@ def _parse_issue_links(self, jira_issue_fields, at_name_jira):
         for issue_link in getattr(jira_issue_fields, at_name_jira):
             if hasattr(issue_link, 'outwardIssue'):
-                issue_id = self._get_issue_id_by_system_id(issue_link.outwardIssue.key)
+                issue_id = issue_link.outwardIssue.key
```
is this not the external_id instead of the issue_id?

myousfi96 replied:
Yes.  I've identified issues with the get_issue_id_by_system_id function in our new design. It would be beneficial for us to discuss and address these problems.

atrautsch commented on issueshark/backends/jirabackend.py:
```diff
@@ -418,16 +409,17 @@ def _parse_issue_links(self, jira_issue_fields, at_name_jira):
                 issue_type, issue_effect = self._get_issue_link_type_and_effect(issue_link.type.outward)
             else:
-                issue_id = self._get_issue_id_by_system_id(issue_link.inwardIssue.key)
+                issue_id = issue_link.inwardIssue.key
```
same as above

atrautsch commented on issueshark/backends/jirabackend.py:
```diff
@@ -565,16 +535,18 @@ def _store_event(self, jira_event, unique_event_id, author_id, created_at, mongo
+        mongo_event = None
+        if mongo_issue:
+            try:
+               mongo_event = IssueEvent.objects(external_id=unique_event_id, issue_id=mongo_issue.id).get()
```
can there be multiple events returned?

atrautsch commented on issueshark/backends/jirabackend.py:
```diff
@@ -586,48 +558,50 @@ def _store_event(self, jira_event, unique_event_id, author_id, created_at, mongo
-                mongo_event.new_value = issue_id
-        elif mongo_event.status == 'issue_links':
+                issue_id = jira_event.toString
+                new_event.new_value = issue_id
```
is this really the issue_id or the external_id or is this resolved later?

myousfi96 commented:
I have updated the test and fixed it in this PR.

myousfi96 commented:
I updated the test to include tests that check our new systems.
I also added a link_issue function that adds Issue_id for linked issues after we store all the issues.

myousfi96 committed:
Refactor issueShark to adapet pycoshark release 3.0 to be able conserve old data

atrautsch commented on issueshark/backends/github.py:
```diff
@@ -106,38 +101,43 @@ def store_issue(self, raw_issue):
-        updated_at = dateutil.parser.parse(raw_issue['updated_at'])
-        created_at = dateutil.parser.parse(raw_issue['created_at'])
-
+        updated_at = dateutil.parser.parse(raw_issue['updated_at']).replace(tzinfo=None)
```
```suggestion
        updated_at = dateutil.parser.parse(raw_issue['updated_at']).astimezone(pytz.UTC).replace(tzinfo=None)
```

atrautsch commented on issueshark/backends/github.py:
```diff
@@ -148,47 +148,38 @@ def _process_events(self, system_id, mongo_issue):
         events_to_store = []
         for raw_event in events:
-            created_at = dateutil.parser.parse(raw_event['created_at'])
+            created_at = dateutil.parser.parse(raw_event['created_at']).replace(tzinfo=None)
```
maybe a short function for this could be in pycoshark so that all plugins can use that.

atrautsch reviewed (changes requested):
Good work, also like the test comments here.
